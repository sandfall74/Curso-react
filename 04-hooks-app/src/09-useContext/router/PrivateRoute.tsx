import { use, type JSX } from "react"
import { UserContext } from "../context/UserContext"
import { Navigate } from "react-router";

interface Props{
    element:JSX.Element,
}

export const PrivateRoute = ({element}:Props) => {
    const {authStatus} = use(UserContext);

    if(authStatus==="checking"){
        return <div>Loading</div>;
    }
    if(authStatus==="authenticated"){
        return element;
    }

  return (
    <Navigate to='/login' replace></Navigate>
  )
}

