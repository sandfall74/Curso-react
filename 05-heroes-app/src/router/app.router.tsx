import { AdminLayout } from "@/admin/layouts/AdminLayout";
import { AdminPage } from "@/admin/pages/AdminPage";
import { HeroesLayout } from "@/heroes/layouts/HeroesLayout";
import { HeroPage } from "@/heroes/pages/hero/HeroPage";
import { HomePage } from "@/heroes/pages/home/HomePage";
import { lazy } from "react";
/* import { SearchPage } from "@/heroes/pages/search/SearchPage"; */
import { createBrowserRouter, Navigate } from "react-router";

const SearchPage=lazy(()=>import('@/heroes/pages/search/SearchPage'));

export const router = createBrowserRouter([
  {
    path: "/",
    element: <HeroesLayout />,
    children: [
      {
        index:true,
        element: <HomePage />,
      },
      {
        path: "heroes/:idSlug",
        element: <HeroPage />,
      },
      {
        path: "search",
        element: <SearchPage />,
      },
      {
        path:'*',
        element:<Navigate to="/"/>
      }
    ],
  },
  {
    path: "/admin",
    element: <AdminLayout />,
    children:[{
        index:true,
        element:<AdminPage/>
    }]
  },
]);
