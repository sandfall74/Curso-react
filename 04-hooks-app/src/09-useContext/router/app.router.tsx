import { createBrowserRouter, Navigate } from "react-router";
import  { AboutPage } from "../pages/About/AboutPage";
import { ProfilePage } from "../pages/profile/ProfilePage";
import { LoginPage } from "../pages/auth/LoginPage";
import { PrivateRoute } from "./PrivateRoute";


export const appRouter = createBrowserRouter([
  {
    path: "/",
    element: <AboutPage/>,
  },{
    path: "/profile",
    element: <PrivateRoute element={<ProfilePage/>}/>,
  },{
    path: "/login",
    element: <LoginPage/>,
  },{
    path: "*",
    element: <Navigate to="/"/>,
  },
]);