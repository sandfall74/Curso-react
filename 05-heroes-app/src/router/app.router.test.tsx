import { describe, expect, test, vi } from 'vitest';
import { router } from './app.router';
import { render, screen } from '@testing-library/react';
import {
  createMemoryRouter,
  Outlet,
  RouterProvider,
  useParams,
} from 'react-router';

vi.mock('@/heroes/layouts/HeroesLayout', () => ({
  HeroesLayout: () => (
    <div data-testid="heroes-layout">
      <Outlet />
    </div>
  ),
}));

vi.mock('@/heroes/pages/home/HomePage', () => ({
  HomePage: () => <div data-testid="home-page"></div>,
}));

vi.mock('@/heroes/pages/hero/HeroPage', () => ({
  HeroPage: () => {
    const { idSlug = '' } = useParams();

    return <div data-testid="hero-page">HeroPage - {idSlug}</div>;
  },
}));

vi.mock('@/heroes/pages/search/SearchPage', () => ({
  default: () => <div data-testid="search-page"></div>,
}));

describe('appRouter', () => {
  test('should be configured as expected', () => {
    expect(router.routes).toMatchSnapshot();
  });

  test('should render home page at root path', () => {
    const testRouter = createMemoryRouter(router.routes, {
      initialEntries: ['/'],
    });
    render(<RouterProvider router={testRouter} />);
    expect(screen.getByTestId('home-page')).toBeDefined();
  });

  test('should render hero page at /heroes/:idSlug path', () => {
    const testRouter = createMemoryRouter(router.routes, {
      initialEntries: ['/heroes/superman'],
    });
    render(<RouterProvider router={testRouter} />);

    expect(screen.getByTestId('hero-page').innerHTML).toContain('superman');
  });

  test('should render search page at /search path', async () => {
    const testRouter = createMemoryRouter(router.routes, {
      initialEntries: ['/search'],
    });

    render(<RouterProvider router={testRouter} />);

    expect(await screen.findByTestId('search-page')).toBeDefined();
  });

  test('should redirect to home page for unknown routes', () => {
    const testRouter = createMemoryRouter(router.routes, {
      initialEntries: ['/otra-pagina-rara'],
    });

    render(<RouterProvider router={testRouter} />);

    expect(screen.getByTestId('home-page')).toBeDefined();
  });
});
