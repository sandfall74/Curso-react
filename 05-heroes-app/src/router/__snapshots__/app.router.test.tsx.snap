// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`appRouter > should be configured as expected 1`] = `
[
  {
    "children": [
      {
        "element": <HomePage />,
        "hasErrorBoundary": false,
        "id": "0-0",
        "index": true,
      },
      {
        "children": undefined,
        "element": <HeroPage />,
        "hasErrorBoundary": false,
        "id": "0-1",
        "path": "heroes/:idSlug",
      },
      {
        "children": undefined,
        "element": <UNDEFINED />,
        "hasErrorBoundary": false,
        "id": "0-2",
        "path": "search",
      },
      {
        "children": undefined,
        "element": <Navigate
          to="/"
        />,
        "hasErrorBoundary": false,
        "id": "0-3",
        "path": "*",
      },
    ],
    "element": <HeroesLayout />,
    "hasErrorBoundary": false,
    "id": "0",
    "path": "/",
  },
  {
    "children": [
      {
        "element": <AdminPage />,
        "hasErrorBoundary": false,
        "id": "1-0",
        "index": true,
      },
    ],
    "element": <AdminLayout />,
    "hasErrorBoundary": false,
    "id": "1",
    "path": "/admin",
  },
]
`;
